javascript/ql/lib/semmle/javascript/security/dataflow/ClientSideUrlRedirectCustomizations.qll

```diff
@@ -124,6 +124,24 @@ module ClientSideUrlRedirect {
     }
   }
 
+  /**
+   * A call to `$("<a>", { href: sink })` or `$(...).attr("href", sink)`.
+   */
+  class JQueryHrefSink extends Sink {
+    JQueryHrefSink() {
+      exists(string prop | prop = DOM::getAPropertyNameInterpretedAsJavaScriptUrl() |
```

Comment:
Good point about `AttributeDefinition`, it already has support for the pattern I was looking for. 

We already used `AttributeDefinition` in the query, but only for a very limited set of attribute names.  
Which is weird, because in `ReactAttributeWriteUrlSink` we just use all the property names from `DOM::getAPropertyNameInterpretedAsJavaScriptUrl()`.  

I'll try to make a `DOMAttributeWriteUrlSink` that behaves similar to `ReactAttributeWriteUrlSink`, and see what the result from that looks like. 

> One of the two new results is of that type and I wouldn't consider that a vulnerability. 

That looks like a missing sanitizer node/edge to me, and not a problem with the sink.  
I can take a look at that after I get results back from  `DOMAttributeWriteUrlSink`.

Reply:
> `AttributeDefinition` should be used here. The jQuery model contributes to that class, and if it's a problem with jQuery it should be a problem regardless of how you assign to the attribute.

And [the new evaluation](https://github.com/github/codeql-dca-main/blob/run/erik-krogh/buildJqueryHref__nightly__CustomSuite__1/reports/alert-comparison.md) looks exactly like the old evaluation.  
I'll look into a sanitizer node/edge.

Reply:
> I'll look into a sanitizer node/edge.

I think the problem is that some sinks are vulnerable to a user-controlled hostname (leading to simple redirection) while others require user-controlled scheme (leading to xss).

I think it's worth looking at a "deep" fix rather than just squeezing an extra sink in here. I think we should try to address these two issues:
- There is no query with dedicated "untrusted URL scheme" sink. This could be the XSS query or a new query.
- The `src` attribute on script tags leads to XSS for untrusted hostnames, but is modelled as client-side url redirection because we can reuse the sanitizer. This is causing confusion however, so it would be nicer if it was in the XSS query.

So I'm wondering if we should add the necessary flow labels to DomBasedXss and just pay the price. We already use `TaintedUrlSuffix` - maybe that can be reused.